compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1005,7 +1053,12 @@ static SizesAndScalableFlags getMatmulVectorSizes(func::FuncOp entryPointFn,
       if (tileSize == 0 || ShapedType::isDynamic(dimSize)) {
         continue;
       }
-
+      // Ad-hoc: Don't attempt to resize scalable tiles when numScalableDims
+      // >= 2. For ArmSME (the only current user of 2D scalable vectors), tile
+      // sizes must match SME tiles (and cannot be arbitrarily resized).
+      if (numScalableDims >= 2 && scalableTileFlags[i]) {
+        continue;
+      }
```

Comment:
I thought the "infra" is more like providing you a way to configure tile sizes. The adjustment is quite independent to the infra. The size adjustment is more about how you compute tile sizes. You can always get the tile sizes you set without adjustment, just don't call those functions?

Anyway, I agree that it is a pain; I've been working progress on removing historical heuristics. I just wanna flag that you will run into the same situation even we have a new tile size selection infra. You will either call the existing methods or re-implement same methods in a different way. The new tile size selection infra does not address the adjustment issue to me. It is an infra, but not a "fix".

Reply:
I think one point we can improve is not having these target-dependent decisions in the generic utilities like `getMatmulVectorSizes` (and [this](https://github.com/openxla/iree/blob/2529fb37c76c89c0306a4bb45f8bdcdf8baeaf16/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1057-L1062)). This can be one reason confusing as we are trying to have a generic tile size selection workflow, but target-specific logic/parameters are segregated in each step.

I'm trying something like dispatching the tile size selection workflow based on the target architecture, then each workflow only needs to take care about its target-specific parameters/logic/adjustment. Then we can factor out common code into target-independent utilities so each workflow can be implemented concisely. Nothing concrete yet but just an idea.